### ISSUE TITLE
Rule Change: Change default for no-unused-var's caughtErrors option to "all"

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [eslint/eslint#17974]
### What rule do you want to change?

no-unused-vars

### What change do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
try {
  // ...
} catch (error) {
  // ...
}
```

### What does the rule currently do for this code?

[`no-unused-var`'s `caughtErrors` option](https://eslint.org/docs/latest/rules/no-unused-vars#caughterrors) defaults to `"none"`. It won't report on an unused caught error by default.

### What will the rule do after it's changed?

With `caughtErrors` defaulting to `"all"`, it would report by default.

Note that this issue just suggests changing the default value for the option. It does not suggest changing any rule logic around naming ignore patterns or removing the option.

### Participation

- [X] I am willing to submit a pull request to implement this change.

### Additional comments

Optional catch bindings were added to ECMAScript 2019 in https://tc39.es/proposal-optional-catch-binding. Per https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch#browser_compatibility they're supported in Node.js >= 10.0.0 and evergreen browsers starting in 2020 (Edge) or <=2018 (others). Given that ESLint's minimum Node version is 12, I think compatibility with older runtimes is no longer a blocker for this issue?

This would be a breaking change. Pity I only thought to request it _after_ the v9 alphas were set in motion 😄. Noticed in https://github.com/typescript-eslint/typescript-eslint/issues/8224.

**Comment by fasttime** [comment]
Makes sense to me, especially because in flat config `ecmaVersion` is `"latest"` by default, and with that setting there is no need for dummy variables bound to a `catch` clause like in ES5. If this change is accepted, I think we are still in time to include it as a breaking change in v9. Looking for more opinions from the team.

**Comment by mdjermanovic** [comment]
I don't think we could change this default, because this rule is in the recommended configuration which should work for everyone. With this change, the recommended configuration couldn't be used out of the box in projects that target environments where the optional catch binding syntax isn't supported.

ESLint not supporting older versions of Node.js means that ESLint cannot be run in those versions, but it doesn't mean that ESLint doesn't support linting code that targets older environments.

**Comment by JoshuaKGoldberg** [comment]
> everyone

Clarifying question: who is "everyone"? Does this include ES3 runtimes?

**Comment by mdjermanovic** [comment]
Yes, a codebase that targets ES3 runtimes for example.

**Comment by JoshuaKGoldberg** [comment]
Got it, thanks. Filed #17975 to suggest making a config that could contain options like this that require newer language targets.

**Comment by bradzacher** [comment]
Drive by 2c: IMO it is wrong to try to prioritise support for codebases writing raw, untranspiled es3. That would be a fraction of a fraction of a fraction of a percent of the total userbase.

IMO you could advance the state of the art for the vast, vast majority of users (those writing ES2019+ in this case) whilst still supporting those outdated users.
For example you could automatically turn the option off by default if the ecmaVersion is `< 2019` and the option isn't explicitly passed by the user.

**Comment by fasttime** [comment]
Maybe we could set the `caughtErrors` option to `false` in the recommended config and still change the default to `true`? This way the recommended config would remain substantially unchanged while the default rule settings would match the preferred presets for new projects.

I can see that all rules in the [recommended config](https://github.com/eslint/eslint/blob/baff28ce8f167f564471f1d70d6e9c4b0cb1a508/packages/js/src/configs/eslint-recommended.js) use the same setting, i.e. `"error"`, so I'm not sure if it's a kind of implicit assumption that rules in the recommended config should use the default options. On the other hand, I don't think there is a general requirement that rule options should allow for maximum compatibility by default. As an example, [`allowKeywords`](https://github.com/eslint/eslint/blob/baff28ce8f167f564471f1d70d6e9c4b0cb1a508/packages/js/src/configs/eslint-recommended.js) in the `dot-notation` rule defaults to `true` at the expenses of ES3 compatibility, and I can totally see why that makes sense considering the way people write code today (e.g. `const foo = (await import("foo")).default`).

In the same vein, I think it should be fine to expect folk to write `try { } catch { }` over `try { } catch (error) { }` nowadays, not necessarily but by default.

**Comment by nzakas** [comment]
We are still in alpha for v9, so we have some latitude for introducing breaking changes. 

I don’t think it’s a good idea to change the recommended config and not the default option. That’s very confusing.

I also don’t think we need to prioritize pre-ES2019 code based when considering defaults. Enabling caughtErrors:all doesn’t have any negative side effects for < ES2019 as it is an unused variable. They can always adjust their options if they don’t want errors reported.

**Comment by nzakas** [comment]
**TSC Summary:** This issue seeks to change the default value of `caughtErrors` on `no-unused-vars` to "all". It's currently "none".

**TSC Question:** Do we want to change this option's default value?

**Comment by sam3k** [comment]
Per 2024-01-25 tsc-meeting, we've agreed to change the default to caughtErrors to "all".

**Comment by snitin315** [comment]
@JoshuaKGoldberg Would you like to submit a PR for this change?

**Comment by JoshuaKGoldberg** [comment]
Can do!
